File: frontend/src/lib/tracking/staerke.ts
// Mannschaftsstärke (Personalstärke) eines Fahrzeugs, geschrieben
// `Führer/Unterführer/Mannschaften//Gesamt` — also etwa `0/1/8//9`.
//
// Einzige Quelle für beide Seiten: den Fahrer-Link, über den die Besatzung
// meldet, und die Tracking-Ansicht, in der die Führung abliest. Die
// Gesamtzahl wird hier gerechnet, nie übertragen — dasselbe gilt hinten in
// `app/services/staerke.py`.

/** Grenze je Rolle, gleichlautend zum Backend (`staerke.MAX_JE_ROLLE`). */
export const MAX_JE_ROLLE = 99;

/** Was in der Anzeige steht, solange nichts gemeldet ist. */
export const OHNE_MELDUNG = '–/–/–';

export interface Staerke {
	fuehrer: number;
	unterfuehrer: number;
	mannschaften: number;
}

/** Die sechs Zahlen, wie Backend-Nutzlasten sie führen. */
export interface StaerkeFelder {
	staerke_soll_fuehrer?: number | null;
	staerke_soll_unterfuehrer?: number | null;
	staerke_soll_mannschaften?: number | null;
	staerke_ist_fuehrer?: number | null;
	staerke_ist_unterfuehrer?: number | null;
	staerke_ist_mannschaften?: number | null;
}

function aus(
	fuehrer: number | null | undefined,
	unterfuehrer: number | null | undefined,
	mannschaften: number | null | undefined,
): Staerke | null {
	// Alle drei leer heißt „nicht angegeben". Eine einzelne Null ist dagegen
	// eine Aussage, deshalb wird hier auf null/undefined geprüft und nicht auf
	// Wahrheitswert — `0 || null` hätte eine Meldung verschluckt.
	if (fuehrer == null && unterfuehrer == null && mannschaften == null) return null;
	return { fuehrer: fuehrer ?? 0, unterfuehrer: unterfuehrer ?? 0, mannschaften: mannschaften ?? 0 };
}

/** Sollstärke aus der Planung, oder `null` wenn keine geplant ist. */
export function sollAus(v: StaerkeFelder): Staerke | null {
	return aus(v.staerke_soll_fuehrer, v.staerke_soll_unterfuehrer, v.staerke_soll_mannschaften);
}

/** Gemeldete Stärke, oder `null` wenn noch nichts gemeldet wurde. */
export function istAus(v: StaerkeFelder): Staerke | null {
	return aus(v.staerke_ist_fuehrer, v.staerke_ist_unterfuehrer, v.staerke_ist_mannschaften);
}

export function gesamt(s: Staerke): number {
	return s.fuehrer + s.unterfuehrer + s.mannschaften;
}

/** `0/1/8//9`, oder `–/–/–` wenn nichts vorliegt. */
export function formatStaerke(s: Staerke | null): string {
	if (!s) return OHNE_MELDUNG;
	return `${s.fuehrer}/${s.unterfuehrer}/${s.mannschaften}//${gesamt(s)}`;
}

/**
 * Weicht die Meldung vom Soll ab?
 *
 * Eine fehlende Meldung ist **keine** Abweichung — unbekannt ist nicht
 * dasselbe wie zu wenig, und die Führung soll die beiden Fälle nicht
 * gleichgesetzt vorfinden.
 */
export function weichtAb(soll: Staerke | null, ist: Staerke | null): boolean {
	if (!soll || !ist) return false;
	return (
		soll.fuehrer !== ist.fuehrer ||
		soll.unterfuehrer !== ist.unterfuehrer ||
		soll.mannschaften !== ist.mannschaften
	);
}

/** Beschriftung für Maus und Screenreader — die Aussage darf nicht allein in der Farbe stecken. */
export function staerkeTitel(soll: Staerke | null, ist: Staerke | null): string {
	if (!ist) {
		return soll
			? `Noch keine Stärkemeldung — Soll ${formatStaerke(soll)}`
			: 'Noch keine Stärkemeldung';
	}
	if (weichtAb(soll, ist)) {
		return `Abweichung vom Soll: gemeldet ${formatStaerke(ist)}, Soll ${formatStaerke(soll)}`;
	}
	return soll
		? `Stärke ${formatStaerke(ist)} — wie geplant`
		: `Stärke ${formatStaerke(ist)}`;
}

/** Gesamtstärke eines Verbands: Summe der Meldungen plus die Zahl der Schweigenden. */
export function verbandsStaerke(fahrzeuge: StaerkeFelder[]): { gemeldet: Staerke; offen: number } {
	const gemeldet: Staerke = { fuehrer: 0, unterfuehrer: 0, mannschaften: 0 };
	let offen = 0;
	for (const f of fahrzeuge) {
		const ist = istAus(f);
		if (!ist) {
			offen++;
			continue;
		}
		gemeldet.fuehrer += ist.fuehrer;
		gemeldet.unterfuehrer += ist.unterfuehrer;
		gemeldet.mannschaften += ist.mannschaften;
	}
	return { gemeldet, offen };
}

File: backend/app/services/staerke.py
"""Mannschaftsstärke (Personalstärke) eines Fahrzeugs im Marschverband.

Geschrieben wird sie ``Führer/Unterführer/Mannschaften//Gesamt``, also etwa
``0/1/8//9``. Gespeichert werden nur die ersten drei Zahlen — die Gesamtzahl
wird gerechnet, nie eingegeben, sonst driftet sie von den Summanden weg.

Hier liegt das Vokabular einmal für alle Wege, auf denen eine Stärke
hereinkommt: der angemeldete Endpunkt und der öffentliche Fahrer-Link. Beide
prüfen dieselben Grenzen, sonst nimmt der eine an, was der andere abweist.
"""

# Mehr als 99 Personen in *einer* Rolle auf *einem* Fahrzeug gibt es nicht; die
# Grenze fängt vertippte Zahlen ab, ohne einen realen Fall auszuschließen.
MAX_JE_ROLLE = 99

_ROLLEN = ("fuehrer", "unterfuehrer", "mannschaften")


def normalisieren(
    fuehrer: int | None,
    unterfuehrer: int | None,
    mannschaften: int | None,
) -> tuple[int, int, int] | None:
    """Die drei Zahlen prüfen und auffüllen.

    Gibt ``None`` zurück, wenn keine einzige Zahl angegeben ist — „nicht
    gemeldet". Ist mindestens eine angegeben, zählen die fehlenden als 0:
    ``0/0/0`` ist dann eine Aussage (niemand an Bord) und kein Fehlen.
    Unplausible Zahlen lösen ``ValueError`` aus.
    """
    werte = (fuehrer, unterfuehrer, mannschaften)
    if all(w is None for w in werte):
        return None

    geprueft: list[int] = []
    for rolle, wert in zip(_ROLLEN, werte):
        if wert is None:
            geprueft.append(0)
            continue
        # bool ist in Python ein int — als Stärke ist es ein Programmfehler.
        if isinstance(wert, bool) or not isinstance(wert, int):
            raise ValueError(f"{rolle}: ganze Zahl erwartet")
        if not 0 <= wert <= MAX_JE_ROLLE:
            raise ValueError(f"{rolle}: muss zwischen 0 und {MAX_JE_ROLLE} liegen")
        geprueft.append(wert)
    return (geprueft[0], geprueft[1], geprueft[2])


def gesamt(
    fuehrer: int | None,
    unterfuehrer: int | None,
    mannschaften: int | None,
) -> int | None:
    """Gesamtstärke, oder ``None`` wenn nichts gemeldet ist.

    Nicht 0 im Fehlen-Fall: die Führung muss „keine Meldung" von „keine
    Besatzung" unterscheiden können.
    """
    werte = normalisieren(fuehrer, unterfuehrer, mannschaften)
    return None if werte is None else sum(werte)
